submissions/iakow/OOP-exercise/index.js

```diff
@@ -0,0 +1,87 @@
+class Inhabitant {
+  constructor({ species, name, gender, legs, hands, say }) {
+    this.species = species;
+    this.name = name;
+    this.gender = gender;
+    this.legs = legs;
+    this.hands = hands;
+    this.say = say;
+    this.friends = [];
+  }
+
+  makeFriends(...args) {
+    const { friends } = this;
+
+    args.forEach((item) => {
+      if (item instanceof Inhabitant && item != this && !friends.includes(item.name)) {
+        friends.push(item.name);
+      }
+    });
+  }
+
+  toString() {
+    const { name, species, gender, legs, hands, say, friends } = this;
+
+    const aboutFriends = (friends.length) ? `My friends: ${friends.join(', ')}.` : '';
+
+    return `${say} About me: ${species}, ${gender}, ${name}, ${hands || 'no'} hands, ${legs} legs. ${aboutFriends}`;
+  }
+}
+
+class Pet extends Inhabitant {
+  constructor({ species, name, gender, say }) {
+    super({ species, name, gender, legs: 4, hands: 0, say });
+  }
+}
+
+class Dog extends Pet {
+  constructor({ name, gender }) {
+    super({ species: 'dog', name, gender, say: 'WOOF-WOOF!!!' });
+  }
+}
+
+class Cat extends Pet {
+  constructor({ name, gender }) {
+    super({ species: 'cat', name, gender, say: 'Meow...' });
+    this.friends = [this.name];
+  }
+}
+
+class Human extends Inhabitant {
+  constructor({ name, gender, say }) {
+    super({ species: 'human', name, gender, legs: 2, hands: 2, say });
+  }
+
+  toString() {
+    const { name, species, gender, legs, hands, say, friends } = this;
+
+    const aboutFriends = (friends.length) ? `I am friends with ${friends.join(', ')}.` : '';
+
+    return `${say} My name is ${name} end I'm ${species}, ${gender}. Obviously, I have ${legs} legs and ${hands} hands.
+${aboutFriends}`;
+  }
+}
+
+class CatWoman extends Human {
+  constructor({ name, say }) {
+    super({ name, gender: 'female', say });
+    this.species = 'humacat';
```

Comment:
It is not a really good pattern to change props owned by a base class in a child class.
Hierarchies can be complex and base class may change the way it handles the property, e.g. it switches to employ `this.constructor.name` to refer to a species, ignoring `species` argument. Humans and pets will still work but this line in `CatWoman` may not reach the desired effect.

Possible solutions:
1. Employ composition pattern
2. Use a method to assign a species inherited from the top base class (which owns this property).

Reply:
I'm not sure if my new solution is correct?

Reply:
And I have a question about the composition pattern. Do it mean creating a mixin, and only that? I have no idea how it might look specifically in my code ... Where can I find examples?